### ISSUE TITLE
Reconnecting Tag Manager with Analytics tag causes error and UX issue

### ISSUE THREAD

**Issue by ernee** [google/site-kit-wp#2513]
## Bug Description

Disconnecting and reconnecting Tag Manager with Analytics Tag configured in the GTM container causes an `Update Settings Failed` error with the following console errors: 

<img width="900" alt="Screen Shot 2020-12-04 at 9 46 19 AM" src="https://user-images.githubusercontent.com/8108643/101690929-cf879380-3a22-11eb-9fbc-619582f44dea.png">

With page refresh, the error clears but the _Confirm & Continue_ button remains disabled creating a blocked experience: 

<img width="1094" alt="Screen Shot 2020-12-04 at 9 36 55 AM" src="https://user-images.githubusercontent.com/8108643/101691198-32792a80-3a23-11eb-8b2b-9043a41373eb.png">

This experience was initially reported [in the support forums](https://wordpress.org/support/topic/cant-connect-to-google-tag-manager-on-site-kit-by-google/). 

Note in my testing the GTM snippet is present in the source code (for non-AMP and AMP pages). I was also able to bypass this issue with either inserting the Analytics snippet (which was previously removed automatically since the GTM Analytics Tag was detected at first connection) before reconnecting Tag Manager or removing the Analytics Tag from the GTM container before reconnecting. 

## Steps to reproduce

1. Connect Analytics and Tag Manager with Analytics Tag in container 
2. Disconnect Tag Manager from Site Kit
3. Reconnect Tag Manager 
4. See `Error: Update Settings Failed` error, refresh and see button `Confirm & Continue` button remains disabled. 

## Additional Context

 - PHP Version: 7.3.23
 - OS: MacOS
 - Browser: Chrome
 - Plugin Version: 1.22.0
 - Device: MacBook Air

Don't think this is the current expected behavior, but noting that the Analytics snippet is not reinserted when Tag Manager with Analytics tag is disconnected or Analytics tag is removed from the GTM container. 

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* No `updating_settings_failed` error should be triggered when reconnecting Tag manager with a matching Analytics tag.
    * This is likely a problem in the `Google\Site_Kit\Core\Storage\Options` class. WordPress's `update_option` and `update_network_option` don't necessarily return `true` if the value to update is exactly the same. However, for our purposes, if the value is the same as the existing one, we should still return `true`.

## Implementation Brief

* Remove custom `Modules::update_module_settings` method (since it no longer does exactly what it says of enforcing all settings are passed)
    - Also, no modules implement a `settings` datapoint anymore and it's better if this works consistently across all modules since the FE expects it to work the same way for all
* Inline the check for `Module_With_Settings` and call `settings->merge( data )` unconditionally
    - **Don't return a `WP_Error` based on the return value of this call** - if it returns `false` it doesn't mean there is a problem, it just means that the value didn't change which is fine
* Merge https://github.com/google/site-kit-wp/compare/fix/2513-module-settings-endpoint?expand=1

### Test Coverage

* None needed

### Visual Regression Changes

* N/A

## QA Brief

On the Site Kit dashboard run this from the console to trigger a settings update that results in no change:
```js
await googlesitekit.data.dispatch( 'modules/search-console' ).saveSettings()
```
* You should get an object returned with a `response` object, and undefined `error`.  
![image](https://user-images.githubusercontent.com/1621608/108328711-98c5ba80-71d5-11eb-9fbe-80c5cd854784.png)
* You should also see the request was made successfully on the network tab  
![image](https://user-images.githubusercontent.com/1621608/108329352-d7f40b80-71d5-11eb-93d3-078af9e7cd51.png)

**Additionally, re-test that GTM setup is no longer affected**

1. Connect Analytics and Tag Manager with Analytics tag in container 
2. Disconnect Tag Manager from Site Kit
3. Reconnect Tag Manager using the same configuration
    - You should no longer see an error or be blocked from completing the setup

## Changelog entry

* Fix bug where in some occasions an error would be displayed when updating settings with the same values that were already stored before.

**Comment by asvinb** [comment]
Hey @aaemnnosttv ,
I tried to reproduce locally but i was unable to do so. Do you think we can reproduce that using the tester plugin?

**Comment by aaemnnosttv** [comment]
@asvinb @felixarntz I'm not sure how this is possible since disconnecting GTM would cause its settings to be deleted. Reconnecting using the same (unchanged) settings in order to see that error would not be possible. I've tried to reproduce this as well and found no problem connecting, disconnecting, and reconnecting GTM using a container that matched the property ID used by Analytics.

**Comment by felixarntz** [comment]
@aaemnnosttv Can you clarify how you got to the IB here? It might be worth going with what you're saying and merge settings instead of expecting all of them to be passed. At the same time though, why are we _not_ passing all settings on the client? Wouldn't that be the root of this JS error, basically what incorrectly used our REST endpoint?

**Comment by aaemnnosttv** [comment]
@felixarntz we already don't require all settings to be passed, since [it uses `Module_Settings::merge` internally](https://github.com/google/site-kit-wp/blob/14cfa1947871bfcdff36c19c68408121d7676bae/includes/Core/Modules/Modules.php#L858-L860) so there's no actual change there. The problem is that we conditionally return a `WP_Error` if `merge` returns `false` which will happen if there is no change to the saved value. In this case `false` just means "not updated" rather than "error" so we shouldn't fail the request. `merge` combines any partial settings value with the current saved settings (but only keeping the known keys) so there's no point in requiring all settings to be passed IMO. I think we would use this method anyways to avoid allowing arbitrary keys to be set in the option.

The source of the problem here is that we update Analytics settings (setting `useSnippet` to `false`) and save them (which will pass them all) except it results in an error because in this case there is no change (snippet was already disabled) so it becomes a failed request.

https://github.com/google/site-kit-wp/blob/14cfa1947871bfcdff36c19c68408121d7676bae/assets/js/modules/tagmanager/components/setup/SetupForm.js#L81-L86

Rather than trying to foresee this error from the client, it's better to let the request go through to the server (since it could be out of sync with the client anyways), but we simply shouldn't give any significance to the return value of the `merge` call. This way the code above stays the same, and is still relevant (e.g. if the save request actually fails for some reason, 500 error, etc).

The rest of it is just cleaning up parts which are no longer used.

**Comment by felixarntz** [comment]
@aaemnnosttv Thanks for clarifying, that makes sense. I wonder though, it feels odd to me (as I already mentioned in the ACs) that not updating the option because it already has that value results in `false` being returned. IMO this is unintuitive: yes there was no update, but that's a low-level detail - the setting _does_ have the value that is expected after the update call. Checking whether an update was successful is somewhat pointless as long as this quirk exists, since the update didn't actually fail in that scenario - it just wasn't necessary. So I think (in addition to your IB) it would be great to address that problem in `Options` / `User_Options`.

**Comment by aaemnnosttv** [comment]
> (in addition to your IB) it would be great to address that problem in `Options` / `User_Options`.

@felixarntz would you please clarify this request? Are you suggesting that we would no longer return anything from the `set` or `delete` methods? I'm not sure we need to change anything with these classes (maybe clarify the documented return value), but perhaps we should do a larger audit to see if these methods are called within any other conditionals in the codebase where a similar problem could happen?

**Comment by felixarntz** [comment]
@aaemnnosttv We should still return a boolean to indicate success/failure, but change it as follows:

* When updating an option to the same value as the current one, no update is necessary. WordPress by default returns `false` because no update was made, but this IMO is misleading because still all is fine (the value that was supposed to be updated to _is_ the value of the option). We shouldn't require the consuming code to always check manually whether no update is necessary. So if no update is necessary, we should return `true`.
* The same applies when deleting an option. If that option is already deleted, WordPress returns `false`. But it already is deleted, so the option is in the expected state. So if no deletion is necessary, we should return `true`.

I guess this is about semantics. Yes no update or deletion is technically made in these cases, so from that angle returning `false` is appropriate. But the expected result of the action is still "successfully" reached in these cases, so I think we should return `true`.

**Comment by aaemnnosttv** [comment]
> I guess this is about semantics. Yes no update or deletion is technically made in these cases, so from that angle returning `false` is appropriate. But the expected result of the action is still "successfully" reached in these cases, so I think we should return `true`.

@felixarntz Yes, but I don't think we should change the return values. The classes we have now are extremely simple and seeing as the return value isn't relevant in most cases (even if we were to make this change) I don't see a good reason to change it.

> We shouldn't require the consuming code to always check manually whether no update is necessary.

I agree, but again, the return value from set/delete methods isn't really used anywhere except in these classes and their tests (I just looked), so I this may have been the only instance where we used the return value with a conditional.

TL;DR  - Let's not change the base methods of the Option classes as part of this issue. Let's save the broader change for another issue, if needed 👍

**Comment by felixarntz** [comment]
IB ✅

**Comment by wpdarren** [comment]
### QA Update: Pass ✅ 

**Verified:**

Ran through the QAB regarding running the command from console

![image](https://user-images.githubusercontent.com/73545194/108724822-b9d53500-7560-11eb-8abf-3a1ea97b54a3.png)
![image](https://user-images.githubusercontent.com/73545194/108724855-bf327f80-7560-11eb-802e-8073ab344045.png)

- No longer see an error or be blocked from completing the setup when Google Analytics tag is included in the container.
